src/datasets.jl

```diff
@@ -335,14 +335,21 @@ write_dataset(dset::Dataset, memtype::Datatype, x, xfer::DatasetTransferProperti
     API.h5d_write(dset, memtype, API.H5S_ALL, API.H5S_ALL, xfer, x)
 
 # type-specific behaviors
-
+function _check_invalid(dataset::Dataset, buf::AbstractArray)
+    num_bytes_dset = Base.checked_mul(sizeof(datatype(dataset)), length(dataset))
+    num_bytes_buf = Base.checked_mul(sizeof(eltype(buf)), length(buf))
+    num_bytes_buf == num_bytes_dset || throw(ArgumentError(
+        "Invalid number of bytes: $(num_bytes_buf) != (num_bytes($(name(dataset))) = $num_bytes_dset)"
```

Comment:
The only thing that's confusing in this error message is `(num_bytes($(name(dataset)))`, since `num_bytes` isn't the name of a valid function. I think you probably meant `sizeof` ?

Reply:
So I tried to use `sizeof` in the beginning, but it was failing on the HDF5 side in the tests (I didn't dig further though).
And for consistency, I also used `sizeof(eltype(...)) * length(...)` on the julia side even though it means `sizeof(buf)`.

Thus, it will be odd to name it `sizeof` even if we don't use `sizeof` under the hood, but I understand that `num_bytes` can be named alternatively even though it is IMO self explanatory.

Maybe @mkitti has an opinion on this ?

Reply:
If we opt for what was suggested the attribute error method should be tweaked to match.

Reply:
I like this. Furthermore, we could argue that `num_bytes` is redundant with the previous sentence element `number of bytes:`.

Reply:
I guess we can leave the attribute error as is, since in that case the definitions agree ?